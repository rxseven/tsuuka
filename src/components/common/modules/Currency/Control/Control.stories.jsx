import React from 'react';
import { action } from '@storybook/addon-actions';
import { storiesOf } from '@storybook/react';

import Control from './index';

const title = 'Components/Common/Modules/Currency/Control';
const props = {
  baseAmount: 1,
  baseCurrency: { code: 'USD', name: 'United States Dollar' },
  baseInputRef: React.createRef(),
  currencies: [
    { code: 'EUR', name: 'Euro' },
    { code: 'THB', name: 'Thai Baht' },
    { code: 'USD', name: 'United States Dollar' }
  ],
  isProcessing: false,
  onAmount: action('onAmount'),
  onExchange: action('onExchange'),
  onReset: action('onReset'),
  onSelect: action('onSelect'),
  override: false,
  rate: '31.7916',
  targetAmount: 0.87892,
  targetCurrency: { code: 'THB', name: 'Thai Baht' },
  targetInputRef: React.createRef()
};

storiesOf(title, module)
  .add('default', () => (
    <Control {...props} rate={null} targetAmount="" targetCurrency={{}} />
  ))
  .add('empty base', () => (
    <Control
      {...props}
      baseCurrency={{}}
      override
      rate={null}
      targetAmount=""
    />
  ))
  .add('processing', () => (
    <Control {...props} baseAmount={1} isProcessing override rate={null} />
  ))
  .add('success', () => <Control {...props} override />)
  .add('amount', () => (
    <Control {...props} baseAmount="100" override targetAmount="87.892" />
  ))
  .add('invalid value', () => (
    <Control
      {...props}
      baseAmount="Invalid value"
      override
      targetAmount="Oops!"
    />
  ));
