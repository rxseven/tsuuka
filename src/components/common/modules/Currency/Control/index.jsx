import PropTypes from 'prop-types';
import exact from 'prop-types-exact';
import { isEmpty } from 'lodash';
import React from 'react';
import styled from 'styled-components';

import Form from 'components/common/base/Form';
import Grid from 'components/common/base/Grid';
import Icon from 'components/common/base/Icon';
import Input from 'components/common/base/Input';
import Label from 'components/common/base/Label';
import Spinner from 'components/common/base/Spinner';
import Combobox from 'components/common/composite/Combobox';
import currencySchema from 'schemas/currency';

import Autocomplete from '../Autocomplete';
import Dropdown from '../Dropdown';

const Base = styled.fieldset`
  padding-right: 0;

  @media (min-width: ${({ theme }) => theme.breakpoint.md}) {
    padding-right: 0.75rem;
  }
`;

const Exchange = styled.div`
  display: flex;
  justify-content: center;
  margin-bottom: 1rem;
  width: 100%;

  @media (min-width: ${({ theme }) => theme.breakpoint.md}) {
    display: block;
    margin-bottom: 0;
    width: auto;
  }
`;

const Flipper = styled(Icon)`
  color: #aaa;
  cursor: pointer;
  font-size: 1.25rem;
  justify-content: center;
  transition: color 0.2s ease-in-out;

  :hover {
    color: ${({ theme }) => theme.color.primary};
  }
`;

const Target = styled.fieldset`
  padding-left: 0;

  @media (min-width: ${({ theme }) => theme.breakpoint.md}) {
    padding-left: 0.75rem;
  }
`;

const Toolbar = styled.div`
  align-items: center;
  display: flex;
  height: 22px;
  justify-content: center;
  width: 25px;

  @media (min-width: ${({ theme }) => theme.breakpoint.md}) {
    left: 50%;
    margin-left: -12.5px;
    position: absolute;
    top: 72px;
    z-index: 1;
  }
`;

// Get suggestion value
// When user navigates the suggestions using the Up and Down keys,
// the input value should be set according to the highlighted suggestion
function getSuggestionValue(suggestion) {
  return suggestion.name;
}

const propTypes = exact({
  baseAmount: PropTypes.oneOfType([PropTypes.number, PropTypes.string])
    .isRequired,
  baseCurrency: PropTypes.oneOfType([
    PropTypes.shape({
      code: PropTypes.string.isRequired,
      name: PropTypes.string.isRequired
    }).isRequired,
    PropTypes.shape({}).isRequired
  ]).isRequired,
  baseInputRef: PropTypes.instanceOf(Object).isRequired,
  currencies: PropTypes.arrayOf(
    PropTypes.shape({
      code: PropTypes.string.isRequired,
      name: PropTypes.string.isRequired
    })
  ).isRequired,
  isProcessing: PropTypes.bool,
  onAmount: PropTypes.func.isRequired,
  onExchange: PropTypes.func.isRequired,
  onReset: PropTypes.func.isRequired,
  onSelect: PropTypes.func.isRequired,
  override: PropTypes.bool,
  rate: PropTypes.string,
  targetAmount: PropTypes.oneOfType([PropTypes.number, PropTypes.string])
    .isRequired,
  targetCurrency: PropTypes.oneOfType([
    PropTypes.shape({
      code: PropTypes.string.isRequired,
      name: PropTypes.string.isRequired
    }).isRequired,
    PropTypes.shape({}).isRequired
  ]).isRequired,
  targetInputRef: PropTypes.instanceOf(Object).isRequired
});

const defaultProps = {
  isProcessing: false,
  override: false,
  rate: null
};

function Control(props) {
  const {
    baseAmount,
    baseCurrency,
    baseInputRef,
    currencies,
    isProcessing,
    onAmount,
    onExchange,
    onReset,
    onSelect,
    override,
    rate,
    targetAmount,
    targetCurrency,
    targetInputRef
  } = props;
  const isReady = !isEmpty(baseCurrency) || !isEmpty(targetCurrency);

  const comboboxProps = {
    data: currencies,
    display: getSuggestionValue,
    dropdown: Dropdown,
    input: Autocomplete,
    override,
    placeholder: 'Enter currency'
  };

  const inputProps = {
    validator: value => {
      if (value && value !== 'Oops!') {
        return currencySchema.isValidSync({ amount: value });
      }

      return true;
    }
  };

  let targetInputlaceholder;

  switch (true) {
    case !isEmpty(targetCurrency) && !isProcessing:
      targetInputlaceholder = 'Amount';
      break;
    case isProcessing:
      targetInputlaceholder = 'Processing...';
      break;
    default:
      targetInputlaceholder = 'Please choose the currency above';
  }

  function handleChangeBase(event) {
    onAmount({ value: event.target.value });
  }

  function handleChangeTarget(event) {
    onAmount({ direction: 'backward', value: event.target.value });
  }

  function handleResetBase() {
    onReset('baseCurrency');
  }

  function handleResetTarget() {
    onReset('targetCurrency');
  }

  function handleSelectBase(value) {
    onSelect({ key: 'baseCurrency', value });
  }

  function handleSelectTarget(value) {
    onSelect({ key: 'targetCurrency', value });
  }

  return (
    <Grid.Row>
      <Grid.Column size="col-md-6">
        <Base>
          <Form.Group>
            <Label htmlFor="combobox-from" small>
              Currency I have
            </Label>
            <Combobox
              {...comboboxProps}
              id="combobox-base"
              initialValue="United States Dollar"
              onReset={handleResetBase}
              onSelect={handleSelectBase}
              value={baseCurrency.name}
            />
          </Form.Group>
          <Form.Group>
            <Label htmlFor="input-base" small>
              I have this much to exchange
            </Label>
            <Input
              {...inputProps}
              name="input-base"
              onChange={handleChangeBase}
              placeholder="Amount"
              ref={baseInputRef}
              value={baseAmount}
            />
          </Form.Group>
        </Base>
      </Grid.Column>
      <Exchange>
        <Toolbar>
          <Choose>
            <When condition={isProcessing}>
              <Spinner circle />
            </When>
            <Otherwise>
              <If condition={isReady}>
                <Flipper
                  icon="retweet"
                  onClick={onExchange}
                  title="Flip currency"
                />
              </If>
            </Otherwise>
          </Choose>
        </Toolbar>
      </Exchange>
      <Grid.Column size="col-md-6">
        <Target>
          <Form.Group>
            <Label htmlFor="combobox-to" small>
              Currency I want
            </Label>
            <Combobox
              {...comboboxProps}
              id="combobox-target"
              onReset={handleResetTarget}
              onSelect={handleSelectTarget}
              value={targetCurrency.name}
            />
          </Form.Group>
          <Form.Group>
            <Label htmlFor="input-target" small>
              I want to buy something at this price
            </Label>
            <Input
              {...inputProps}
              disabled={!rate}
              name="input-target"
              onChange={handleChangeTarget}
              placeholder={targetInputlaceholder}
              ref={targetInputRef}
              value={rate ? targetAmount : ''}
            />
          </Form.Group>
        </Target>
      </Grid.Column>
    </Grid.Row>
  );
}

Control.propTypes = propTypes;
Control.defaultProps = defaultProps;

export default Control;
