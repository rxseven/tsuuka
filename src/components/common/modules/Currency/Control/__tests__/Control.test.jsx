import { shallow } from 'enzyme';
import React from 'react';
import renderer from 'react-test-renderer';

import Environment from 'tests/environment';
import Control from '../index';

describe('<Currency.Control />', () => {
  // Arrange
  const props = {
    baseAmount: 100,
    baseCurrency: { code: 'USD', name: 'United States Dollar' },
    baseInputRef: React.createRef(),
    currencies: [
      { code: 'EUR', name: 'Euro' },
      { code: 'THB', name: 'Thai Baht' },
      { code: 'USD', name: 'United States Dollar' }
    ],
    isProcessing: false,
    onAmount: jest.fn(),
    onExchange: jest.fn(),
    onReset: jest.fn(),
    onSelect: jest.fn(),
    override: false,
    rate: '31.7916',
    targetAmount: '87.892',
    targetCurrency: { code: 'THB', name: 'Thai Baht' },
    targetInputRef: React.createRef()
  };
  const component = (
    <Environment>
      <Control {...props} />
    </Environment>
  );

  describe('Unit tests', () => {
    it('should render without crashing', () => {
      // Act
      const wrapper = shallow(component);

      // Assert
      expect(wrapper).toBeDefined();
    });
  });

  describe('Snapshot tests', () => {
    it('should render correctly', () => {
      // Act
      const tree = renderer.create(component).toJSON();

      // Assert
      expect(tree).toMatchSnapshot();
    });
  });
});
