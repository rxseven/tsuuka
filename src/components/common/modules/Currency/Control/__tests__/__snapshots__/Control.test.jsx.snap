// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Currency.Control /> Snapshot tests should render correctly 1`] = `
<div
  className="row"
>
  <div
    className="col-md-6"
  >
    <fieldset
      className="sc-gzVnrw cZywgI"
    >
      <div
        className="form-group"
      >
        <label
          className="sc-bdVaJa eOvgbZ"
          htmlFor="combobox-from"
        >
          Currency I have
        </label>
        <div
          aria-expanded={false}
          aria-haspopup="listbox"
          aria-owns="react-autowhatever-combobox-base"
          className="react-autosuggest__container"
          role="combobox"
        >
          <div
            className="sc-bwzfXH ghuPdm"
          >
            <input
              aria-activedescendant={null}
              aria-autocomplete="list"
              aria-controls="react-autowhatever-combobox-base"
              autoComplete="off"
              className="form-control"
              id="combobox-base"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              placeholder="Enter currency"
              type="text"
              value="United States Dollar"
            />
            <svg
              aria-hidden="true"
              className="svg-inline--fa fa-times-circle fa-w-16 sc-htpNat kLyKcn"
              data-icon="times-circle"
              data-prefix="fas"
              focusable="false"
              onClick={[Function]}
              role="img"
              style={Object {}}
              viewBox="0 0 512 512"
              visibility="United States Dollar"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M256 8C119 8 8 119 8 256s111 248 248 248 248-111 248-248S393 8 256 8zm121.6 313.1c4.7 4.7 4.7 12.3 0 17L338 377.6c-4.7 4.7-12.3 4.7-17 0L256 312l-65.1 65.6c-4.7 4.7-12.3 4.7-17 0L134.4 338c-4.7-4.7-4.7-12.3 0-17l65.6-65-65.6-65.1c-4.7-4.7-4.7-12.3 0-17l39.6-39.6c4.7-4.7 12.3-4.7 17 0l65 65.7 65.1-65.6c4.7-4.7 12.3-4.7 17 0l39.6 39.6c4.7 4.7 4.7 12.3 0 17L312 256l65.6 65.1z"
                fill="currentColor"
                style={Object {}}
              />
            </svg>
          </div>
          <div
            className="react-autosuggest__suggestions-container"
            id="react-autowhatever-combobox-base"
            role="listbox"
          />
        </div>
      </div>
      <div
        className="form-group"
      >
        <label
          className="sc-bdVaJa eOvgbZ"
          htmlFor="input-base"
        >
          I have this much to exchange
        </label>
        <input
          className="form-control"
          name="input-base"
          onChange={[Function]}
          placeholder="Amount"
          type="text"
          value={100}
        />
      </div>
    </fieldset>
  </div>
  <div
    className="sc-htoDjs cSitJf"
  >
    <div
      className="sc-gZMcBi bEfcsw"
    >
      <svg
        aria-labelledby="svg-inline--fa-title-Bs4REF0LDroH"
        className="svg-inline--fa fa-retweet fa-w-20 sc-dnqmqq hLJzAU"
        data-icon="retweet"
        data-prefix="fas"
        onClick={[MockFunction]}
        role="img"
        style={Object {}}
        viewBox="0 0 640 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <title
          id="svg-inline--fa-title-Bs4REF0LDroH"
          style={Object {}}
        >
          Flip currency
        </title>
        <path
          d="M629.657 343.598L528.971 444.284c-9.373 9.372-24.568 9.372-33.941 0L394.343 343.598c-9.373-9.373-9.373-24.569 0-33.941l10.823-10.823c9.562-9.562 25.133-9.34 34.419.492L480 342.118V160H292.451a24.005 24.005 0 0 1-16.971-7.029l-16-16C244.361 121.851 255.069 96 276.451 96H520c13.255 0 24 10.745 24 24v222.118l40.416-42.792c9.285-9.831 24.856-10.054 34.419-.492l10.823 10.823c9.372 9.372 9.372 24.569-.001 33.941zm-265.138 15.431A23.999 23.999 0 0 0 347.548 352H160V169.881l40.416 42.792c9.286 9.831 24.856 10.054 34.419.491l10.822-10.822c9.373-9.373 9.373-24.569 0-33.941L144.971 67.716c-9.373-9.373-24.569-9.373-33.941 0L10.343 168.402c-9.373 9.373-9.373 24.569 0 33.941l10.822 10.822c9.562 9.562 25.133 9.34 34.419-.491L96 169.881V392c0 13.255 10.745 24 24 24h243.549c21.382 0 32.09-25.851 16.971-40.971l-16.001-16z"
          fill="currentColor"
          style={Object {}}
        />
      </svg>
    </div>
  </div>
  <div
    className="col-md-6"
  >
    <fieldset
      className="sc-iwsKbI CkFnx"
    >
      <div
        className="form-group"
      >
        <label
          className="sc-bdVaJa eOvgbZ"
          htmlFor="combobox-to"
        >
          Currency I want
        </label>
        <div
          aria-expanded={false}
          aria-haspopup="listbox"
          aria-owns="react-autowhatever-combobox-target"
          className="react-autosuggest__container"
          role="combobox"
        >
          <div
            className="sc-bwzfXH ghuPdm"
          >
            <input
              aria-activedescendant={null}
              aria-autocomplete="list"
              aria-controls="react-autowhatever-combobox-target"
              autoComplete="off"
              className="form-control"
              id="combobox-target"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              placeholder="Enter currency"
              type="text"
              value=""
            />
            <svg
              aria-hidden="true"
              className="svg-inline--fa fa-times-circle fa-w-16 sc-htpNat fTQKpC"
              data-icon="times-circle"
              data-prefix="fas"
              focusable="false"
              onClick={[Function]}
              role="img"
              style={Object {}}
              viewBox="0 0 512 512"
              visibility=""
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M256 8C119 8 8 119 8 256s111 248 248 248 248-111 248-248S393 8 256 8zm121.6 313.1c4.7 4.7 4.7 12.3 0 17L338 377.6c-4.7 4.7-12.3 4.7-17 0L256 312l-65.1 65.6c-4.7 4.7-12.3 4.7-17 0L134.4 338c-4.7-4.7-4.7-12.3 0-17l65.6-65-65.6-65.1c-4.7-4.7-4.7-12.3 0-17l39.6-39.6c4.7-4.7 12.3-4.7 17 0l65 65.7 65.1-65.6c4.7-4.7 12.3-4.7 17 0l39.6 39.6c4.7 4.7 4.7 12.3 0 17L312 256l65.6 65.1z"
                fill="currentColor"
                style={Object {}}
              />
            </svg>
          </div>
          <div
            className="react-autosuggest__suggestions-container"
            id="react-autowhatever-combobox-target"
            role="listbox"
          />
        </div>
      </div>
      <div
        className="form-group"
      >
        <label
          className="sc-bdVaJa eOvgbZ"
          htmlFor="input-target"
        >
          I want to buy something at this price
        </label>
        <input
          className="form-control"
          disabled={false}
          name="input-target"
          onChange={[Function]}
          placeholder="Amount"
          type="text"
          value="87.892"
        />
      </div>
    </fieldset>
  </div>
</div>
`;
